packages/react/src/useFetch.ts

```diff
@@ -0,0 +1,146 @@
+import { Reducer, useCallback, useEffect, useReducer, useRef } from "react";
+import { useConnection } from "./GadgetProvider";
+import { useStructuralMemo } from "./useStructuralMemo";
+import { ErrorWrapper } from "./utils";
+
+export interface FetchHookState<T> {
+  data?: T;
+  response?: Response;
+  error?: ErrorWrapper;
+  fetching: boolean;
+  options: FetchHookOptions;
+}
+
+export type FetchHookResult<T> = [FetchHookState<T>, (opts?: Partial<FetchHookOptions>) => Promise<T>];
+
+type FetchAction<T> = { type: "fetching" } | { type: "fetched"; payload: T } | { type: "error"; payload: ErrorWrapper };
+
+const reducer = <T>(state: FetchHookState<T>, action: FetchAction<T>): FetchHookState<T> => {
+  switch (action.type) {
+    case "fetching":
+      return { ...state, fetching: true, error: undefined };
```

Comment:
If `state.fetching`, should we return state instead of an object with a new identity? Similarly for others (but may need to check all of the params).

Reply:
🤷 I think its ok to return an object with a new identity if a dispatch has happened -- they only really happen when something momentous really does happen. It's not perfectly optimal but it makes this code a lot simpler?

Reply:
Up to you! The dispatch works similar to some of the other state stuff in that same identity won't cause a re-render. From https://react.dev/reference/react/useReducer#dispatch.

> If the new value you provide is identical to the current state, as determined by an [Object.is](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) comparison, React will skip re-rendering the component and its children.